### PR TITLE
Doc IndexLinear and IndexCartesian

### PR DESCRIPTION
**Pull request by kshyatt**
Hopefully this is ok as a summary ... definitely open to suggestions for improvement!

### CODE REVIEW

**Review by timholy — COMMENTED** [review]
On my phone. Looks great, thanks. One nuance: you can index any array with either style---these are just signals from the developer about which one is implemented/faster.

File: base/indices.jl [hunk 1]
```diff
@@ -13,7 +13,30 @@ Indices{N} = NTuple{N,AbstractUnitRange}
 ## Traits for array types ##
 
 abstract type IndexStyle end
+"""
+    IndexLinear()
+
+Subtype of [`IndexStyle`](@ref) which specifies that the array it describes
```
Comment (mbauman) [inline base/indices.jl]: Really nice.  Here's a shot at a set of hopefully-slightly-clearer first sentences:

> Subtype of ``[`IndexStyle`](@ref)`` that is used to describe arrays that are optimally indexed by one **linear** index.

> Subtype of ``[`IndexStyle`](@ref)`` that is used to describe arrays that are optimally indexed by a **Cartesian** index with the same number of indices as there are dimensions in the array.

I also have a slight preference for separating out the first sentence in a docstring as a stand-alone "summary" line (a paragraph by itself).

**Comment by kshyatt** [comment]
Thanks for the feedback, guys. I'll get on this asap.

File: base/indices.jl [hunk 2]
```diff
@@ -13,22 +13,47 @@ Indices{N} = NTuple{N,AbstractUnitRange}
...
 """
     IndexStyle(A)
     IndexStyle(typeof(A))
 
 `IndexStyle` specifies the "native indexing style" for array `A`. When
-you define a new `AbstractArray` type, you can choose to implement
-either linear indexing or cartesian indexing.  If you decide to
-implement linear indexing, then you must set this trait for your array
+you define a new [`AbstractArray`](@ref) type, you can choose to implement
+either linear indexing (with [`IndexLinear`](@ref)) wor cartesian indexing.
```
Comment (andyferris) [inline base/indices.jl]: “wor cartesian” should be “or Cartesian” (w + capital)

**Review by andyferris — COMMENTED** [review]
Is it specified how offset arrays should transform between linear and Cartesian indexes? The description seems accurate for 1-based arrays.

File: base/indices.jl (hunk 2 above)
Comment (kshyatt) [reply base/indices.jl]: `wor` vs `xor` ... 😛

**Comment by kshyatt** [comment]
Good to go? Whoever merges, please remember to remove the `[ci skip]` when you squash!

**Pull request merged**